server/controllers/finance/reports/ohada_balance_sheet/index.js

```diff
@@ -0,0 +1,299 @@
+        }
+
+        if (item.ref === 'DF') {
+          list = ['DA', 'DB', 'DC', 'DD'];
+          _.extend(item, aggregateReferences(list, currentReferences, previousReferences));
+        }
+
+        if (item.ref === 'DG') {
+          list = ['CL', 'CM', 'DA', 'DB', 'DC', 'DD'];
+          _.extend(item, aggregateReferences(list, currentReferences, previousReferences));
+        }
+
+        if (item.ref === 'DP') {
+          list = ['DH', 'DI', 'DJ', 'DK', 'DL', 'DM', 'DN'];
+          _.extend(item, aggregateReferences(list, currentReferences, previousReferences));
+        }
+
+        if (item.ref === 'DT') {
+          list = ['DQ', 'DR', 'DS'];
+          _.extend(item, aggregateReferences(list, currentReferences, previousReferences));
+        }
+
+        if (item.ref === 'DZ') {
+          list = ['CL', 'CM', 'DA', 'DB', 'DC', 'DD']
+            .concat(['DH', 'DI', 'DJ', 'DK', 'DL', 'DM', 'DN'])
+            .concat(['DQ', 'DR', 'DS'])
+            .concat(['DV']);
+          _.extend(item, aggregateReferences(list, currentReferences, previousReferences));
+        }
+        return item;
+      });
+
+      _.merge(context, { assetTable, liabilityTable });
+      return report.render(context);
+    })
+    .then(result => {
+      res.set(result.header).send(result.report);
+    })
+    .catch(next);
+}
+
+function aggregateReferences(references, currentDb, previousDb) {
+  const item = {
+    currentBrut : 0, currentAmo : 0, currentNet : 0, previousNet : 0,
+  };
+
+  references.forEach(ref => {
+    item.currentBrut += currentDb[ref] ? currentDb[ref].brut.balance : 0;
+    item.currentAmo += currentDb[ref] ? currentDb[ref].amortissement.balance : 0;
+    item.currentNet += currentDb[ref] ? currentDb[ref].net.balance : 0;
+    item.previousNet += previousDb[ref] ? previousDb[ref].net.balance : 0;
+  });
+
+  return item;
+}
+
+function formatReferences(references) {
+  const values = {};
+  Object.keys(references).forEach(key => {
+    const [brut] = references[key].filter(elt => elt.is_amo_dep === 0);
+    let [amortissement] = references[key].filter(elt => elt.is_amo_dep === 1);
+
+    if (!amortissement) {
+      amortissement = { balance : 0 };
+    }
+
+    const net = {
+      abbr : brut.abbr,
+      description : brut.description,
+      balance : brut.balance - amortissement.balance,
+    };
+
+    values[key] = { brut, amortissement, net };
+  });
+  return values;
+}
+
+function getFiscalYearDetails(fiscalYearId) {
+  const bundle = {};
+  // get fiscal year details and the last period id of the fiscal year
+  const query = `
+    SELECT 
+      p.id AS period_id, p.end_date,
+      fy.id, fy.label, fy.previous_fiscal_year_id 
+    FROM fiscal_year fy 
+    JOIN period p ON p.fiscal_year_id = fy.id 
+      AND p.number = (SELECT MAX(period.number) FROM period WHERE period.fiscal_year_id = ?)
+    WHERE fy.id = ?;
+  `;
+  return db.one(query, [fiscalYearId, fiscalYearId])
```

Comment:
I think this would be easier to understand if you just did two calls to `getFiscalYearDetails()`.  That way, the function does what we expect - just gets the fiscal year info for one fiscal year identified by the `fiscalYearId`.  It also will make it easier to change in the future if we need to factor in multiple years or something.

The pseudocode would look like this:

```js
let bundle = {};
getFiscalYearDetails(params.fiscal_id)
  .then(fiscalYear => {
    bundle.current = fiscalYear;
    
    return fiscalYear.previous_fiscal_year_id ? getFiscalYearDetails(fiscalYear.previous_fiscal_year_id) : {};
  })
  .then(previousFiscalYear => {
   // etc...
  });
```

What do you think?